mantidproject/mantid
# Support rotations and translations in LoadSampleShape

**Description of work.**

Add support for translations and rotations into LoadSampleShape algorithm based on support that already exists in LoadSampleEnvironment. Note that these rotations are distinct from the rotations applied by the goniometer. The new rotations added in this PR are to eliminate any unwanted rotations\offsets in the .stl file. The CAD software that generated the file may not have the mesh centred on the origin for example.

As well as the main changes in LoadSampleShape.cpp, the work involved creating a new base class called LoadSingleMesh that is the parent of LoadStl and LoadOff classes so can do the rotations and translations in concise way across all both formats supported in LoadSampleShape (.stl in binary and ascii forms and .off). The local "rotate" function in LoadSampleShape was removed as part of this because it duplicated code already present in the LoadXXX class hierarchy.

**To test:**

1) Save this .stl file [ScrewLowResolution.zip](https://github.com/mantidproject/mantid/files/4845209/ScrewLowResolution.zip) so it's in one of your Mantid data search directories (or change path in test script below)
2) Have a look at the stl file in a CAD application that allows you to view the shape and provides read out of coordinates. I couldn't find an online viewer that provided the coordinate readout unfortunately so you'll probably need to install something. I've been using FreeCAD (https://www.freecadweb.org/) . Notice that the screw points in the y direction and that the centre of the screw is at ~(0,55,63)
3) then run this script which loads the .stl file, rotates the screw so it lies along the z axis and offsets it so it is centred on the origin (the rotation is applied first so the y\z offsets don't match the coordinates of the centre I described in step 2):
```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

ws = CreateSampleWorkspace()
ws = LoadSampleShape(InputWorkspace=ws,
                   Filename=r'ScrewLowResolution.stl',
                   Scale='cm',
                   XDegrees=90,
                   TranslationVector='0,63,-55')
SaveSampleEnvironmentAndShape(InputWorkspace=ws,
                                Filename=r'ScrewLowResolutionRotatedAndCentred.stl',
                                Scale='cm')
```
4) view the xxxRotatedAndCentred.stl file in the CAD application to check it has indeed been rotated\translated

Fixes #28728 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
